src/main/scala/analysis/RegToMemAnalysis.scala

```diff
@@ -0,0 +1,67 @@
+package analysis
+
+import ir.{MemoryLoad, *}
+import analysis.solvers.*
+
+import scala.collection.immutable
+
+/**
+ * Collects all the memory loads and the expressions that are assigned to a register but cannot be evaluated.
+ *
+ * Tracks:
+ * R_x = MemoryLoad[Base + Offset]
+ * R_x = Base + Offset
+ *
+ * Both in which constant propagation mark as TOP which is not useful.
+ */
+trait RegionAccessesAnalysis(cfg: ProgramCfg, constantProp: Map[CfgNode, Map[Variable, FlatElement[BitVecLiteral]]]) {
+
+  val mapLattice: MapLattice[RegisterVariableWrapper, FlatElement[Expr], FlatLattice[Expr]] = MapLattice(FlatLattice[_root_.ir.Expr]())
+
+  val lattice: MapLattice[CfgNode, Map[RegisterVariableWrapper, FlatElement[Expr]], MapLattice[RegisterVariableWrapper, FlatElement[Expr], FlatLattice[Expr]]] = MapLattice(mapLattice)
+
+  val domain: Set[CfgNode] = cfg.nodes.toSet
+
+  val first: Set[CfgNode] = Set(cfg.startNode)
+
+  /** Default implementation of eval.
+   */
+  def eval(cmd: Command, constants:  Map[Variable, FlatElement[BitVecLiteral]], s: Map[RegisterVariableWrapper, FlatElement[Expr]]): Map[RegisterVariableWrapper, FlatElement[Expr]] = {
+    cmd match {
+      case localAssign: LocalAssign =>
+        localAssign.rhs match {
+          case memoryLoad: MemoryLoad =>
+            s + (RegisterVariableWrapper(localAssign.lhs) -> FlatEl(memoryLoad).asInstanceOf[FlatElement[Expr]])
+          case binaryExpr: BinaryExpr =>
+            if (evaluateExpression(binaryExpr.arg1, constants).isEmpty) { // approximates Base + Offset
+              return s + (RegisterVariableWrapper(localAssign.lhs) -> FlatEl(binaryExpr).asInstanceOf[FlatElement[Expr]])
+            }
```

Comment:
How does this approximate 'base + offset'? evaluateExpression can return None for plenty of other cases that aren't just aren't handled at present, which will just cause cascading problems. This isn't robust at all.

Why do we even care about the concept of 'base + offset' - that is a heuristic which does not apply to ARM64.

Reply:
@yousifpatti

Reply:
The heuristic of a base + offset is essential for the working out of regions for the MemoryModelMap. While it may not work perfectly for adrp instructions, it still provide a good approximation of the memory layout. What is an alternative approach that you suggest that does not depend on base + offset for stack regions?

The way the approximation happens is by assuming that the base pointer (ie. lhs in a binaryExpr) is an abstract location that cannot be evaluated and as such if it does not evaluate to anything at that point of the program then it may be a base pointer and the lhs is an offset.

Reply:
"can't be evaluated" is currently defined very broadly since evaluateExpression doesn't support many operators. It also isn't even checking that it is an addition or subtraction operation?

Reply:
adrp instructions are also the most common way that pointers to the data section are loaded from memory, so it isn't good if we don't handle those well.

Reply:
> We don't need to be so pessimistic, it still seems relevant in a lot of [cases](https://nationally-usable-bull.ngrok-free.app/z/KE7Y8f), and makes sense as the way to symbolically treat addresses. Yes its not a heuristic we can naively apply everywhere and we should be more structured about it. But, e.g. `adrp` is not a huge deal because we usually only want the final address value anyway to find a global variable or block address.
> 
> It would be good to more thoroughly look into what heuristics we can use and where, if you have suggestions on how to infer structure from access patterns please share them. Ultimately these types of analysis (VSA/DSA/ASI) need heuristics like this to identify how to group memory accesses into regions and the ARM load/store addressing mode specification should give us something to work with.

What I would suggest is just making better use of the information in the symbol table - that already gives us what the MemoryRegionAnalysis is attempting to infer. The symbol table directly gives us the addresses of objects in the data region. These objects can then be divided up further according to accesses that take place within them.

Reply:
Currently, the Data Regions are loaded straight from the symbol table and divided based on their addresses. However, we do not have information about stack regions that we can acquire from an external source so the heuristic is an ok approximation for those regions for our purposes

Reply:
Not all the information that could be retrieved from the symbol table is retrieved currently, so there is definitely room for improvement there.

It isn't really a heuristic for the stack accesses - it is more just a case of ensuring that the value of the stack pointer is tracked effectively so that offsets to it are evaluated correctly. This part is not really an issue.